waspc/data/Generator/templates/react-app/src/router/linkHelpers.ts

```diff
@@ -0,0 +1,49 @@
+export type ParamValue = string | number
+export type Params = { [name: string]: ParamValue }
+export type Search =
+  | string[][]
+  | Record<string, string>
+  | string
+  | URLSearchParams
+
+export function interpolatePath(
+  path: string,
+  params?: Params,
+  search?: Search,
+  hash?: string
+) {
+  const interpolatedPath = params ? interpolatePathParams(path, params) : path
+  const interpolatedSearch = search
+    ? `?${new URLSearchParams(search).toString()}`
+    : ''
+  const interpolatedHash = hash ? `#${hash}` : ''
+  return interpolatedPath + interpolatedSearch + interpolatedHash
+}
+
+function interpolatePathParams(path: string, params: Params) {
+  const startsWithSlash = path.startsWith("/");
+  const interpolatedPath = path
+    .split("/")
+    .map((part) => {
+      if (part.startsWith(":")) {
+        const paramName = getParamName(part);
+        return params[paramName];
+      }
+      return part;
+    })
+    // Drops empty strings
```

Comment:
I recommend removing all the comments explaining what the code does (especially for non-complex stuff).

If you think the code is unclear and would benefit from an explanation, it's better to extract it into a named function or assign the complex expression to a named variable.

Great video on the topic: https://www.youtube.com/watch?v=Bf7vDBBOBUA